File: apps/mp/api/payment.ts
/**
 * 支付 API 客户端 (FE-S3-004).
 *
 * 对齐后端 ``apps/api/app/api/v1/payment.py`` (BE-S3-010):
 * - POST ``/api/v1/pay/wechat/order``  下单 (auth required, 10/min/user 限流)
 * - POST ``/api/v1/pay/wechat/notify`` 微信回调 (前端不调; 微信服务器直推)
 *
 * 字段 100% 对齐 ``apps/api/app/schemas/payment.py``:
 * - ``CreateOrderRequest``     ``{plan, payment_channel}``
 * - ``CreateOrderResponse``    ``{order_id, out_trade_no, plan, amount_cny, payment_channel, payment_params, created_at}``
 * - ``PaymentParams``          ``{timeStamp, nonceStr, package, signType, paySign}`` — 微信 JSAPI 协议硬约束 mixedCase
 *
 * 设计取舍:
 *
 * - **不做字段映射 / 不做客户端价目表**: 价格走后端 ``PLAN_PRICES_CNY`` 单一权威表,
 *   前端 UI 展示金额从 ``CreateOrderResponse.amount_cny`` 取 (实际服务端把套餐价目映射成金额).
 *   防"前端价目表与后端不同步"导致用户看到 ¥99 但实付 ¥299
 *
 * - **``payment_params`` 直接喂 ``uni.requestPayment``**: 5 字段 (timeStamp / nonceStr /
 *   package / signType / paySign) 与 ``wx.requestPayment`` 入参 1:1 对齐, 前端零字段
 *   映射. 这是和 BE-S3-010 schema 设计一起规划的 — schema 层故意不 snake_case 化
 *
 * - **``payment_channel`` 默认 ``wechat_mp``**: 当前仅 MP-WEIXIN 走真实支付; H5 / App
 *   下单时 BE 也接受 ``wechat_mp`` 但 SDK 层会失败 (没 wx_openid). 跨端守卫由 composable
 *   层 ``gotoPay()`` 兜底, 不让 H5 / App 误调
 *
 * 错误码 (后端 ``HTTPException(detail={'code','message'})``):
 * - 400 ``invalid_plan``               套餐非法 (实际不会到这, schema validate 422)
 * - 400 ``wechat_openid_required``     用户没 wx_openid (手机号注册用户在生产环境需先 wx.login)
 * - 422 unprocessable                  schema 校验失败 (plan / payment_channel 不匹配 Literal)
 * - 429 ``rate_limit_exceeded``        10/min/user 限流
 * - 502 ``sdk_error``                  微信 SDK 调用失败 (上游故障)
 */

import { request } from '@/utils/request'

/** 真实付费档; 试用走后端 vip_service.grant_trial, 不走此路径 */
export type PayablePlan = 'monthly' | 'quarterly' | 'yearly' | 'lifetime'

/** 支付渠道; 当前仅 ``wechat_mp`` (小程序内 JSAPI) 走真实下单 */
export type PaymentChannel = 'wechat_mp' | 'wechat_h5'

export interface CreateOrderRequest {
  plan: PayablePlan
  /** 默认 ``wechat_mp``, 服务端默认值同步 */
  payment_channel?: PaymentChannel
}

/**
 * 微信 JSAPI 5 字段; ``uni.requestPayment`` 入参原样透传.
 *
 * **mixedCase 字段名是微信硬协议**: ``wx.requestPayment`` / ``uni.requestPayment``
 * 都要求字段名严格驼峰, 这里和 BE Pydantic schema (``# noqa: N815``) 一致, 不做
 * snake_case 转换.
 */
export interface PaymentParams {
  timeStamp: string
  nonceStr: string
  /** 形如 ``prepay_id=wx2026042811...``, 来自 BE SDK 调用结果 */
  package: string
  signType: 'RSA'
  /** 商户私钥 RSA-SHA256 签名 (Base64); 微信侧用商户公钥验签 */
  paySign: string
}

export interface CreateOrderResponse {
  order_id: string
  out_trade_no: string
  plan: PayablePlan
  /** 服务端权威金额 (CNY); 前端展示 / 对账用, 不重算 */
  amount_cny: number
  payment_channel: PaymentChannel
  payment_params: PaymentParams
  created_at: string
}

/**
 * 创建微信支付订单 (BE-S3-010 ``POST /pay/wechat/order``).
 *
 * 登录态强制; 限流 10/min/user. 同 ``user + plan + payment_channel`` 在 5 min 窗口内
 * 重复下单复用旧 pending 订单, 防双击 + 网络重试 + 用户慌乱反复点.
 *
 * 调用方 (composable / vip page) 拿到 ``payment_params`` 后调:
 *
 * ```ts
 * uni.requestPayment({
 *   provider: 'wxpay',
 *   ...resp.payment_params,
 *   success: () => navigateTo('/pages/vip/result?status=paid&order_id=' + resp.order_id),
 *   fail: (err) => { ... },
 * })
 * ```
 */
export function createWechatOrder(req: CreateOrderRequest) {
  return request<CreateOrderResponse>({
    url: '/api/v1/pay/wechat/order',
    method: 'POST',
    data: { payment_channel: 'wechat_mp', ...req },
  })
}

File: apps/api/app/schemas/payment.py
"""微信支付 v3 域 Pydantic 模型 (BE-S3-010).

下单 / 回调请求 / 响应 4 类:

- ``CreateOrderRequest``    ← ``POST /pay/wechat/order``  请求体
- ``CreateOrderResponse``   ← ``POST /pay/wechat/order``  响应
- ``PaymentParams``         ← ``CreateOrderResponse.payment_params`` 内嵌; 5 件套字段 100% 对齐
  ``uni.requestPayment`` 入参 (前端不做字段映射)
- ``NotifyResponse``        ← ``POST /pay/wechat/notify`` 响应; 微信 v3 协议固定结构

错误响应统一走 ``HTTPException(detail={'code': '...', 'message': '...'})``,
不在 schema 层定义错误模型 (与 auth / agent 模块一致).
"""

from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

# 仅接受真实付费档; 试用走 vip_service.grant_trial 不走此路径
PayablePlanLiteral = Literal["monthly", "quarterly", "yearly", "lifetime"]
PaymentChannelLiteral = Literal["wechat_mp", "wechat_h5"]


class CreateOrderRequest(BaseModel):
    """``POST /pay/wechat/order`` 请求体.

    - ``plan``: 必填; 仅 4 种付费档. 服务端自带价目表算金额, 前端不传 amount (防篡改).
    - ``payment_channel``: 默认 ``wechat_mp`` (小程序 JSAPI). H5 / Native 走不同 SDK 调用.
    """

    model_config = ConfigDict(extra="forbid")

    plan: PayablePlanLiteral
    payment_channel: PaymentChannelLiteral = Field(
        default="wechat_mp",
        description="支付渠道; 当前仅支持 wechat_mp (小程序内 JSAPI). wechat_h5 预留 Sprint 4+",
    )


class PaymentParams(BaseModel):
    """``uni.requestPayment`` 调用入参 5 件套, 微信小程序 JSAPI 协议固定.

    与 ``wx.requestPayment`` 1:1 对齐, 字段名严格用驼峰 (微信协议要求, 前端零字段映射).

    N815 mixedCase 是微信硬协议要求 (``wx.requestPayment`` 入参字段名固定),
    本项目其它地方走 snake_case; 这里独立例外, 走 noqa.
    """

    model_config = ConfigDict(extra="forbid")

    timeStamp: str = Field(  # noqa: N815  WeChat Pay JSAPI protocol field name
        description="Unix 秒时间戳字符串 (10 位); 与 paySign 计算输入一致"
    )
    nonceStr: str = Field(  # noqa: N815  WeChat Pay JSAPI protocol field name
        description="随机串 (≤ 32 位); 与 paySign 计算输入一致"
    )
    package: str = Field(description="prepay_id 包装串, 格式 'prepay_id=wx...'")
    signType: Literal["RSA"] = Field(  # noqa: N815  WeChat Pay JSAPI protocol field name
        default="RSA", description="v3 协议固定 RSA-SHA256"
    )
    paySign: str = Field(  # noqa: N815  WeChat Pay JSAPI protocol field name
        description="商户私钥签名结果 (Base64); 微信侧用商户公钥验签后授权拉起支付"
    )


class CreateOrderResponse(BaseModel):
    """``POST /pay/wechat/order`` 响应.

    - ``order_id`` / ``out_trade_no``: 内部业务键 + 商户订单号; 前端持久化用
    - ``amount_cny``: 计价确认 (前端展示用; 与请求 plan 对应的服务端价目表一致)
    - ``payment_params``: 直接喂给 ``uni.requestPayment``, 前端无需做任何变换
    """

    model_config = ConfigDict(extra="forbid")

    order_id: uuid.UUID
    out_trade_no: str
    plan: PayablePlanLiteral
    amount_cny: Decimal
    payment_channel: PaymentChannelLiteral
    payment_params: PaymentParams
    created_at: datetime


class NotifyResponse(BaseModel):
    """``POST /pay/wechat/notify`` 响应; 微信 v3 协议固定结构.

    协议要求即使业务失败也返回 HTTP 200 + body ``{code: 'FAIL', message: '...'}``;
    返非 200 / 非 SUCCESS 时微信会持续重试 24h (按指数退避), 容易把订单状态搞乱.

    成功固定: ``{code: 'SUCCESS', message: 'OK'}``.
    """

    model_config = ConfigDict(extra="forbid")

    code: Literal["SUCCESS", "FAIL"]
    message: str = Field(default="OK", max_length=128)


__all__ = [
    "CreateOrderRequest",
    "CreateOrderResponse",
    "NotifyResponse",
    "PayablePlanLiteral",
    "PaymentChannelLiteral",
    "PaymentParams",
]
